Libplanet/Net/Swarm.cs

```diff
@@ -2003,21 +2033,44 @@ private void TransferRecentStates(GetRecentStates getRecentStates)
                 {
                     Guid chainId = BlockChain.Id;
 
+                    _logger.Debug(
+                        "Getting state references from {Offset}",
+                        getRecentStates.Offset);
+
+                    long lowestIndex =
+                        (@base is HashDigest<SHA256> bbh &&
+                         _store.GetBlockIndex(bbh) is long bbIdx)
+                            ? bbIdx + getRecentStates.Offset
+                            : 0 + getRecentStates.Offset;
+                    long targetIndex =
+                        (target is HashDigest<SHA256> tgt &&
+                         _store.GetBlockIndex(tgt) is long tgtIdx)
+                            ? tgtIdx
+                            : long.MaxValue;
+                    var highestIndex = lowestIndex + FindNextStatesChunkSize - 1 > targetIndex
+                        ? targetIndex
+                        : lowestIndex + FindNextStatesChunkSize - 1;
+
+                    nextOffset = highestIndex == targetIndex
+                        ? -1
+                        : getRecentStates.Offset + FindNextStatesChunkSize;
+
                     stateRefs = _store.ListAllStateReferences(
                         chainId,
-                        onlyAfter: @base,
-                        ignoreAfter: target
+                        lowestIndex: lowestIndex,
+                        highestIndex: highestIndex
                     );
 
+                    // GetBlockStates may return null since swarm may not have deep states.
                     blockStates = stateRefs.Values
                         .Select(refs => refs.Last())
                         .ToImmutableHashSet()
+                        .Where(bh => !(_store.GetBlockStates(bh) is null))
                         .Select(bh =>
                             new KeyValuePair<
                                 HashDigest<SHA256>,
                                 IImmutableDictionary<Address, IValue>
-                            >(bh, _store.GetBlockStates(bh))
-                        )
+                            >(bh, _store.GetBlockStates(bh)))
```

Comment:
I guess it's better if we call `_store.GetBlockStates()` once and filter by its value.